File: LeetCodeProblemsCMake/Problems/LeetCode1774ClosestDessertCost.cpp
//
// Created by saeed on 9/23/25.
//

#include "LeetCode1774ClosestDessertCost.h"

#include <iostream>
#include <numeric>
#include <bits/ranges_algo.h>


using namespace std;



int LeetCode1774ClosestDessertCost::closestCost(vector<int> &baseCosts, vector<int> &toppingCosts, int target) {
    int maxBase=*ranges::max_element(baseCosts);
    int sumToppings=accumulate(toppingCosts.begin(), toppingCosts.end(), 0)*2;
    int maxPrice=maxBase+sumToppings;
    vector<bool> canMake(maxPrice+1,false);
    for (auto  const base: baseCosts) {
        canMake[base] = true;
    }

}

void LeetCode1774ClosestDessertCost::Run() {
    vector<int> baseCosts{3,10};
    vector<int> toppings{2,5};
    int target{9};
    cout<<closestCost(baseCosts,toppings,target)<<endl;
}

File: LeetCodeProblemsGo/LeetCode1774ClosestDessertCost/LeetCode1774ClosestDessertCost.go
package LeetCode1774ClosestDessertCost

import (
	"math"
	"slices"
)

func closestCost(baseCosts []int, toppingCosts []int, target int) int {
	maxPrice := 0
	maxPrice = slices.Max(baseCosts)
	for _, topping := range toppingCosts {
		maxPrice += topping * 2
	}
	canMake := make([]bool, maxPrice+1)
	for _, base := range baseCosts {
		canMake[base] = true
	}
	for _, topping := range toppingCosts {
		for i := 0; i < 2; i++ {
			for j := maxPrice; j >= topping; j-- {
				if canMake[j-topping] {
					canMake[j] = true
				}
			}
		}
	}
	closesetDelta := math.MaxInt32
	closest := 0
	var delta int
	for i := 0; i <= maxPrice; i++ {
		if !canMake[i] {
			continue
		}
		delta = i - target
		if delta < 0 {
			delta *= -1
		}
		if closesetDelta > delta {
			closesetDelta = delta
			closest = i
		}
	}
	return closest
}
